clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -796,6 +796,126 @@ test_synthetic()
+test_speculation()
+{
+    std::cerr << "\n----------------\nTesting speculation\n";
+    std::vector<trace_entry_t> memrefs = {
+        /* clang-format off */
+        make_thread(1),
+        make_pid(1),
+        make_marker(TRACE_MARKER_TYPE_PAGE_SIZE, 4096),
+        make_timestamp(10),
+        make_marker(TRACE_MARKER_TYPE_CPU_ID, 1),
+        // Conditional branch.
+        make_instr(1, TRACE_TYPE_INSTR_CONDITIONAL_JUMP),
+        // It fell through in the trace.
+        make_instr(2),
+        // Another conditional branch.
+        make_instr(3, TRACE_TYPE_INSTR_CONDITIONAL_JUMP),
+        // It fell through in the trace.
+        make_instr(4),
+        make_instr(5),
+        make_exit(1),
+        /* clang-format on */
+    };
+    std::vector<scheduler_t::input_reader_t> readers;
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(memrefs)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), 1);
+
+    scheduler_t scheduler;
+    std::vector<scheduler_t::input_workload_t> sched_inputs;
+    sched_inputs.emplace_back(std::move(readers));
+    scheduler_t::scheduler_options_t sched_ops =
+        scheduler_t::make_scheduler_serial_options(/*verbosity=*/4);
+    sched_ops.flags = static_cast<scheduler_t::scheduler_flags_t>(
+        static_cast<int>(sched_ops.flags) |
+        static_cast<int>(scheduler_t::SCHEDULER_SPECULATE_NOPS));
+    if (scheduler.init(sched_inputs, 1, sched_ops) != scheduler_t::STATUS_SUCCESS)
+        assert(false);
+    int ordinal = 0;
+    auto *stream = scheduler.get_stream(0);
+    memref_t memref;
+    for (scheduler_t::stream_status_t status = stream->next_record(memref);
+         status != scheduler_t::STATUS_EOF; status = stream->next_record(memref)) {
+        assert(status == scheduler_t::STATUS_OK);
+        switch (ordinal) {
+        case 0:
+            assert(memref.marker.type == TRACE_TYPE_MARKER);
+            assert(memref.marker.marker_type == TRACE_MARKER_TYPE_PAGE_SIZE);
+            break;
+        case 1:
+            assert(memref.marker.type == TRACE_TYPE_MARKER);
+            assert(memref.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP);
+            break;
+        case 2:
+            assert(memref.marker.type == TRACE_TYPE_MARKER);
+            assert(memref.marker.marker_type == TRACE_MARKER_TYPE_CPU_ID);
+            break;
+        case 3:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 1);
+            break;
+        case 4:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 2);
+            // We realize now that we mispredicted that the branch would be taken.
+            // We ask to queue this record for post-speculation.
+            stream->start_speculation(100, true);
+            break;
+        case 5:
+            // We should now see nops from the speculator.
+            assert(type_is_instr(memref.instr.type));
```

Comment:
This doesn't verify that it's a nop. Maybe we can have a hard-coded check here for the encoding.

Reply:
I tried linking in the decoder to decode and run instr_is_nop() and that worked w/o too much effort so I'll go with that.

Reply:
Maybe pulling in the decoder was a mistake: dealing with Windows toolchain issues due to that...